.github/workflows/ci.yml

```diff
@@ -0,0 +1,192 @@
+name: Build and test
+
+# The main C++ CI is split into three steps: lint, build, and system tests. The
+# linting and building are independent but we make the build wait so as not to
+# waste resources compiling and testing a commit which doesn't pass the linter.
+#
+# System tests are slow and need input data. To save caching, network, and job
+# time we only run the full system tests on Ubuntu (although this can easily be
+# changed).
+
+
+on:
+  push:
+    branches: ["main"]
+  pull_request:
+  workflow_dispatch:
+
+
+concurrency:
+  # Skip intermediate builds: always.
+  # Cancel intermediate builds: always.
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+
+jobs:
+  linting:
+    name: Linting checks with pre-commit
+    runs-on: ubuntu-latest
+    steps:
+      - uses: ucl/composite-actions/pre-commit@v1
+
+  build:
+    name: ${{ matrix.build_type }} build on ${{ matrix.os }} with build testing ${{ matrix.build_testing }}
+    needs: linting
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-latest, windows-latest, macos-latest]
+        build_type: [Debug]
+        # Only Debug builds (so we have helpful debug output in our logs). Can
+        # also test 'Release'.
+        build_testing: [ON, OFF]
+        exclude:
+          # Currently our tests don't compile in Windows so skip this.
+          - os: windows-latest
+            build_testing: ON
+
+    steps:
+      - name: Check out repository
+        uses: actions/checkout@v3
+
+      - name: Set up MATLAB
+        uses: matlab-actions/setup-matlab@v1.2.3
+
+      # -------------------------------------------------------------------------------
+      # Ubuntu
+      - name: Install dependencies for Ubuntu
+        if: ${{ contains(matrix.os, 'ubuntu') }}
+        run: |
+          sudo apt-get update
+          sudo apt-get install libfftw3-dev libgomp1 python3
+
+      # -------------------------------------------------------------------------------
+      # Windows
+      - name: Install miniconda for Windows
+        if: ${{ contains(matrix.os, 'windows') }}
+        uses: conda-incubator/setup-miniconda@v2
+        with:
+          auto-update-conda: true
+          python-version: 3.9
```

Comment:
Just making a note for myself here, but `MATLABEngine` (which the test-regeneration currently uses) is only compatible with `Python >=3.10`. Although given we don't run the tests on Windows, :shrug: